keyserver/listener.go

```diff
@@ -0,0 +1,109 @@
+package keyserver
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net"
+	"time"
+
+	. "github.com/coniks-sys/coniks-go/protocol"
+)
+
+func (server *ConiksServer) listenForRequests(ln net.Listener, handler func(msg []byte) ([]byte, error)) {
+	defer ln.Close()
+	go func() {
+		<-server.stop
+		server.waitCloseConn.Wait()
+		ln.Close()
+	}()
+
+	for {
+		conn, err := ln.Accept()
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
```

Comment:
This return invokes the deferred close though, and even if you remove that, it clears the wait group on return.